### PR TITLE
Let cupertino & material switches move to the right state after dragging

### LINKED ISSUE #46046: [Widget] Unexpected drag behaviour on Switch
When a Switch widgets value is determined via *a function of type Future**, the drag behaviour overrides the value of the Switch(visually).

*: This will be referred as __toggleSwitch_()

If __toggleSwitch_() completes with _false_ but the _onChanged_ was called after the switch was toggled on by dragging, the switch's value will be displayed as _true_. This is only a visual bug, hot reload won't fix the issue. 

You would expect the switch's value to visually reset after __toggleSwitch_() completes with _false_ in this case.

I'm not sure of the expected behaviour because of this description:
```
/// The switch itself does not maintain any state. Instead, when the state of
/// the switch changes, the widget calls the [onChanged] callback. Most widgets
/// that use a switch will listen for the [onChanged] callback and rebuild the
/// switch with a new [value] to update the visual appearance of the switch.
```

image
## Steps to Reproduce
[**Application**](https://gist.githubusercontent.com/erdemyerebasmaz/154dc46cca97a0c52621dcbd6eba9756/raw/36351826e1b303fa9b797ef10d80168ebd48545f/main.dart)
In this application there's a Switch with a non-null _onChanged_ property. A function of type Future called __toggleSwitch_() is called when this Switch is toggled on. __toggleSwitch_() sets the Switch to a random _bool_ value after 100ms. This function is only called when the user toggles the switch on, the switch resets to _false_ when toggled off.

1. Test tap gestures behaviour to verify the Switch's value and visual indicator is in-line.
2. Restart the app
3. Toggle switch by dragging till __toggleSwitch_() completes with _false_.
   1. Reset either by dragging or tapping if __toggleSwitch_() completes with _true_.
   2. If __toggleSwitch_() completes with _false_, tapping / hot reload won't fix this visual bug. You would need to toggle off by dragging before continuing, so the value and visual indicator is in-line.

I think this can also be replicated the other way around; where the switch is toggled of by dragging and __toggleSwitch_() completes with _true_.

**Target Platform: Android
**Target OS version/browser: Android 9
**Devices: Device(Haven't tested with Emulator)

## Logs
[**flutter run -v**](https://gist.githubusercontent.com/erdemyerebasmaz/ab4fd667e14846b28ff0570804d83509/raw/6ee4e573b11c4822ad7b9ea07915d28caac164d8/gistfile1.txt)

**flutter analyze**
```
Analyzing switch_drag_test...
No issues found! (ran in 1.9s)
```

**flutter doctor -v**
```
[√] Flutter (Channel beta, v1.11.0, on Microsoft Windows [Version 10.0.18362.476], locale en-GB)
    • Flutter version 1.11.0 at C:\Development\sdk\flutter
    • Framework revision 856a90e67c (4 weeks ago), 2019-11-08 18:00:01 -0800
    • Engine revision af04338413
    • Dart version 2.7.0

[√] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at C:\Development\sdk\Android
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • ANDROID_HOME = C:\Development\sdk\Android
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b03)
    • All Android licenses accepted.

[√] Android Studio (version 3.5)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 41.1.2
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b03)

[√] VS Code (version 1.38.1)
    • VS Code at C:\Users\erdem\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 2.22.2

[√] Connected device (1 available)
    • EML L09 • BPN7N18419008822 • android-arm64 • Android 9 (API 28)

• No issues found!
```
_(It was 5:30 am when I wrote this. I'm happy to rephrase parts that are incomprehensible.)_

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
## Description

This pull request makes Cupertino/Material switches move their thumbs to the correct side after a drag gesture.  The `onChanged` callback is called when the drag gesture ends at a position that indicates the switch should change its value (which is consistent with the onTap behavior, making them both call `onChanged` when the user reveals the intent to change the value. However this may be a breaking change). 

For example, with the below switch:
```dart
Switch(
  value: false,
  onChanged: (bool newValue) { },
);
``` 
If the thumb is dragged to `true` (which is still doable), it will animate back to `false` when the user lets go.

In addition to that, it also moves `CupertinoSwitch`'s gestures/animation logic to its state from the render object.

## Related Issues

Fixes #46046

## Tests

I added the following tests:

- "can veto switch dragging result" for both switches.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change. Does not seem to break internal tests.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by Afsar-Pasha** [comment]
The switch bounces a little bit
t3
Looks like its caused by changing `_position`'s curves in the middle.

**Comment by LongCatIsLooong** [comment]
@Afsar-Pasha thanks for pointing that out! I disabled curve switching for explicit animations (those not triggered by `widget.value` changes). Turns out there's an issue for a similar problem: https://github.com/flutter/flutter/issues/51627.

**Comment by Afsar-Pasha** [comment]
LGTM
Can you also fix material switch, I tried but it looks like the logic can't be moved from render object to state as `_RenderSwitch` extends `RenderToggleable` which contains most of the logic.

Also the current behavior is a little different `RenderToggleable`(or current `CupertinoSwitch`) calls `onChanged`  whenever `AnimationStatus.completed` or `AnimationStatus.dismissed`(dragging switch to the end will call `onChange` even without lifting the finger up) but this does not.
I prefer this behavior than the old one, and we can't make material switch to behave like this while extending `RenderToggleable`.

**Comment by LongCatIsLooong** [comment]
I'll start working on this today.

**Review by HansMuller — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/cupertino/switch.dart [hunk 1]
```diff
@@ -140,8 +140,166 @@ class CupertinoSwitch extends StatefulWidget {
...
+  TapGestureRecognizer _tap;
+  HorizontalDragGestureRecognizer _drag;
+
+  AnimationController _positionController;
+  CurvedAnimation position;
+
+  AnimationController _reactionController;
+  Animation<double> _reaction;
+
+  TextDirection textDirection;
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/cupertino/switch.dart]: removed

**Pull request merged**